storage/Devices/DeviceImpl.h

```diff
@@ -21,6 +21,32 @@ namespace storage
     class SystemInfo;
 
 
+    template <typename Type>
+    Type* to_device_of_type(Device* device)
+    {
+	static_assert(!is_const<Type>::value, "Type must not be const");
+
+	Type* tmp = dynamic_cast<Type*>(device);
+	if (!tmp)
+	    ST_THROW(DeviceHasWrongType("device has wrong type"));
```

Comment:
Making the exception class a templated class visible in the public header files does not look like a good idea to me. That would likely cause trouble with some or all binding target languages.

But the two class names can be passed to the exception. Still, traits are needed and that is a bit more work.